Add smpp-connector-example.md (#1537)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/smpp-connector/smpp-connector-example.md

# SMPP Connector Example

> SMPP (Short Message Peer to Peer protocol) Connector allows you to send SMS through the WSO2 EI. It uses [jsmpp API](https://jsmpp.org/) to communicate with a SMSC (Short Message service center) which is to store, forward, convert and deliver Short Message Service (SMS) messages. jsmpp is a java implementation of SMPP protocol. 

Reviewer: ```suggestion
SMPP (Short Message Peer-to-Peer Protocol) Connector allows you to send an SMS through the WSO2 EI. It uses the [jsmpp API](https://jsmpp.org/) to communicate with an SMSC (Short Message Service Center), which is used to store, forward, convert, and deliver Short Message Service (SMS) messages. jsmpp is a Java implementation of the SMPP protocol. 
```

Author: Updated.

## What you'll build

> Given below is a sample scenario that demonstrates how to work WSO2 EI SMPP Connector and send SMS messages via the SMPP protocol.

Reviewer: ```suggestion
Given below is a sample scenario that demonstrates how to work with the WSO2 EI SMPP Connector and send SMS messages via the SMPP protocol.
```

Author: Updated.

> The SMPP server in SMSC have all ESME (External Short Messaging Entity) addresses which is an external application that connects to a SMSC and the active connection. When you send SMS to destination, it comes to the SMSC, then one of the module in SMSC checks the destination address whether it is available or not. If it is available it create connection object that is responsible for sending SMS message.

Reviewer: ```suggestion
The SMPP server in SMSC have all the ESME (External Short Messaging Entity) addresses. This is an external application that connects to a SMSC and the active connection. When you send an SMS to a destination, it comes to the SMSC. Then one of the modules in SMSC checks if the destination address is available or not. If it is available, it creates a connection object that is responsible for sending the SMS message.
```

Author: Updated.

> There are many SMPP gateways available in the world.and now almost all the Message centers support SMPP. It is not practical always to connect with real SMSC. However in this scenario we will try it with [SMSC simulator](https://github.com/smn/logica-smpp-sim).

Reviewer: ```suggestion
There are many SMPP gateways available in the world and now almost all the message centers support SMPP. It is not practical always to connect with real SMSC. However, in this scenario we will try it with [SMSC simulator](https://github.com/smn/logica-smpp-sim).
```

Author: Updated as following manner.
There are many SMPP gateways available in the world and now almost all the message centers support SMPP. It is not practical always to connect with real SMSC. However, in this scenario we will try it with **SMSC simulator**. Please refer the [Setting up SMPP connector](setting-up-smpp-connector.md) documentation.

The following `sendSMS`operation is exposed via an API. The API with the context `/send` has one resource.

> * `/send` : Use to send SMS Message to the Short Message service center.

Reviewer: ```suggestion
* `/send` : Used to send SMS Message to the Short Message Service Center.
```

Author: Updated.

> The following diagram shows the overall solution. There is HTTP API and user can invoke it with http call with JSON. API is able to send a SMS for the request number in JSON request with message in JSON. 

Reviewer: ```suggestion
The following diagram shows the overall solution. There is an HTTP API that you can invoke with an HTTP call with JSON. The API is able to send a SMS for the request number in a JSON request with the message in JSON. 
```

Author: Updated.

<img src="../../../../assets/img/connectors/smpp-connector-example.png" title="SMPP connector example" width="800" alt="smpp connector example"/>

If you do not want to configure this yourself, you can simply [get the project](#get-the-project) and run it.

## Configure the connector in WSO2 Integration Studio

> Follow these steps to set up the ESB Solution Project and the Connector Exporter Project.

Reviewer: ```suggestion
Follow these steps to set up the Integration Project and the Connector Exporter Project.
```

Author: Updated.

### Import the connector

Follow these steps to set up the ESB Solution Project and the Connector Exporter Project. 

{!references/connectors/importing-connector-to-integration-studio.md!} 

### Add integration logic

First create an API, which will be where we configure the integration logic. Right click on the created Integration Project and select, **New** -> **Rest API** to create the REST API. Specify the API name as `SmppTestApi` and API context as `/send`.
    
<img src="../../../../assets/img/connectors/adding-an-api.png" title="Adding a Rest API" width="800" alt="Adding a Rest API"/>

#### Configuring the API

Create a resource to send an SMS to the Short Message Service Center.

1. Initialize the connector.
    
    1. Navigate into the **Palette** pane and select the graphical operations icons listed under **SMPP Connector** section. Then drag and drop the `init` operation into the Design pane.
        
        <img src="../../../../assets/img/connectors/smpp-drag-and-drop-init.png" title="Drag and drop init operation" width="500" alt="Drag and drop init operation"/>   
    

>     2. Add the property values into the `init` operation as shown bellow. Replace the `host`, `port`, `systemId`, `password` with your values.

Reviewer: ```suggestion
    2. Add the property values into the `init` operation as shown below. Replace the `host`, `port`, `systemId`, `password` with your values.
```

Author: Updated.

           
        <img src="../../../../assets/img/connectors/smpp-drag-and-drop-sendSMS.png" title="Drag and drop send operation" width="500" alt="Drag and drop send operation"/>    

    2. In this operation we are going to send a SMS messages peer to peer using SMPP protocol. It provides a flexible data communications interface for transfer of short message data between a Message Centers, such as a Short Message Service Centre (SMSC), GSM Unstructured Supplementary Services Data (USSD) Server or other type of Message Center and a SMS application system, such as a WAP Proxy Server, EMail Gateway or other Messaging Gateway. Please find the `send` operation parameters listed here.
               
        - **sourceAddress** : Source address of the SMS message. 
        - **destinationAddress** : Destination address of the SMS message. 
        - **message** : Content of the SMS message.
        
        While invoking the API, the above three parameters values come as a user input.
        
        <img src="../../../../assets/img/connectors/smpp-drag-and-drop-sendSMS-parameters.png" title="Drag and drop send operation" width="500" alt="Drag and drop send operation"/> 
    

>     3. To get the input values in to the API we can use the [property mediator](https://ei.docs.wso2.com/en/next/micro-integrator/references/mediators/property-Mediator/). Navigate into the **Palette** pane and select the graphical mediators icons listed under **Mediators** section. Then drag and drop the `Property` mediators into the Design pane as shown bellow.

Reviewer: Please use a relative link here.

Author: Updated.

   
        <img src="../../../../assets/img/connectors/smpp-api-property-mediator-property1-value1.png" title="Add property mediators sourceAddress" width="600" alt="Add property mediators sourceAddress"/>
    
    5. Add the property mediator to capture the `message` values. The message contains content of the SMS message.                  
   
        - **name** : message
        - **expression** : json-eval($.message)
     
        <img src="../../../../assets/img/connectors/smpp-api-property-mediator-property2-value2.png" title="Add values to capture message" width="600" alt="Add values to capture message"/>  
      
    6. Add the property mediator to capture the `distinationAddress` values. The message contains content of the SMS message.                  
       
        - **name** : distinationAddress
        - **expression** : json-eval($.distinationAddress)
         
        <img src="../../../../assets/img/connectors/smpp-api-property-mediator-property3-value3.png" title="Add values to capture distinationAddress" width="600" alt="Add values to capture distinationAddress"/>  
        
3. Get a response from the user.
    

>     When you invoking the created API the request of the message is going through the `/send` resource. Finally pass to the the [Respond mediator](https://ei.docs.wso2.com/en/next/micro-integrator/references/mediators/respond-Mediator/). In here the Respond Mediator stops the processing on the current message and sends the message back to the client as a response.            

Reviewer: ```suggestion
    When you are invoking the created API, the request of the message is going through the `/send` resource. Finally, it is passed to the [Respond mediator](https://ei.docs.wso2.com/en/next/micro-integrator/references/mediators/respond-Mediator/). The Respond Mediator stops the processing on the current message and sends the message back to the client as a response.            
```

Add a relative link here.

Author: Updated.

    
    1. Drag and drop **respond mediator** to the **Design view**. 
    
         <img src="../../../../assets/img/connectors/smpp-drag-and-drop-respond-mediator.png" title="Add Respond mediator" width="800" alt="Add Respond mediator"/> 

    2. Once you have setup the sequences and API, you can see the `salesforcerest` API as shown below.
    
         <img src="../../../../assets/img/connectors/smpp-api-design-view.png" title="API Design view" width="600" alt="API Design view"/>
        

>     > **Note**: That the properties should be add to the pallet before create the operation.

Reviewer: ```suggestion
    > **Note**: The properties should be added to the pallet before creating the operation.
```

Author: Updated.
